docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,573 @@
+*   have no *inherent* side effects to the running container image
+
+## Debug Patterns
+
+There are two patterns that are useful when debugging a pod:
+
+1.  *Copy Debug* creates a copy of a pod with minor changes to the pod spec
+    specified on the command or complex changes specified using an editor.
+1.  *Running Debug* mode causes Kubernetes to run a new *debug container* in the
+    pod context. A *debug container* is not part of the pod spec and has only
+    limited configuration by the user. It is created by a debug operation on an
+    existing pod and reported in `PodStatus`.
+
+### Copy Debug
+
+*Copy Debug* works by fetching the spec of a running pod, modifying it based on
+command line arguments, and creating a new pod. This fits well within the
+existing `kubectl run` command, which exists to create new pods.
+
+An example use might be to modify the entrypoint of a pod that's crash looping:
+
+```
+kubectl run target-pod-copy --copy-of=target-pod -it --attach --container=crashing-container --command -- sh
+```
+
+Command line arguments allow simple, single-container changes but are
+insufficient for tasks like modifying volumes and their mount points. For more
+complex changes, we can provide an `--edit` workflow to adjust the generated
+spec before creation:
+
+```
+# Similar to the following workflow:
+#   kubectl get -o yaml ... > $temp_file
+#   $EDITOR $temp_file
+#   kubectl create -f $temp_file
+kubectl run --copy-of=target-pod -it -c shell --image=debian --edit
+```
+
+Note that the `--edit` workflow is useful for all generated config, not just
+copies of pods.
+
+#### Scheduling conflicts
+
+Attempting to duplicate some resources will cause the resulting pod to be
+unschedulable, for example creating a second read-write volume mount of a
+gcePersistentDisk will create a disk conflict. This is the same problem faced by
+scaling a deployment, so one way to provide a better user experience is to
+validate the generated config similar to `ValidatePodTemplateSpecForRC()` with
+replicas > 1 as `kubectl scale` does.
+
+Validating client side is not optimal, and currently only gcePersistentDisk
+provides a check in `ValidatePodTemplateSpecForRC()`, but this isn't really a
+new problem introduced by this new functionality.
+
+#### Stripping labels
+
+Only the spec will be copied from the running pod. Copying metadata such as
+labels might result in the pod copy receiving traffic or being killed
+immediately by a replication controller. It would be trivial to provide a
+`--copy-labels` option if this is desired behavior.
+
+### Running Debug
+
+*Running Debug* requires new functionality in the kubelet and a new `kubectl
+debug` command that results in a new container being introduced in the running
+pod. This *debug container* is not part of the pod spec, which remains
+immutable, and is not restarted according to the pod's `restartPolicy`.
+
+The status of a *debug container* is reported in a new `DebugContainerStatuses`
```

Comment:
If there are multiple active debug containers, how could the user tell which one is created by them? Do they rely on the "name" of the containers to figure that out? If that's the case, since you will not GC the terminated debug containers, users can never reuse the same name. I think that's ok, but this is not explained clearly in the proposal.